File: public/js/raytracing/BresenhamRoute.js
var BresenhamRoute = function BresenhamRoute(start, end, tileRepository)
{
    this.start = start;
    this.end = end;
    this.tileRepository = tileRepository;
};

BresenhamRoute.prototype.getRoute = function()
{
    return this.bresenhamRaytrace(this.start, this.end);
};

BresenhamRoute.prototype.getRouteWithDetails = function()
{
    if ( ! this.tileRepository)
        throw "You need to give a valid tileRepository (TileGrid)";

    var route = this.bresenhamRaytrace(this.start, this.end);

    var newRoute = [];

    route.every(function(step)
    {
        if (step.isArray)
        {
            var tileA = this.tileRepository.getGameTileByXY(step[1].x, step[1].y);
            var tileB = this.tileRepository.getGameTileByXY(step[2].x, step[2].y);

            //if either corner is out of the map, ignore.
            if ( ! tileA || ! tileB)
                return true;

            newRoute.push([this.getDetailsFromTile(tileA), this.getDetailsFromTile(tileA)]);
        }
        else
        {
            var tile = this.tileRepository.getGameTileByXY(step.x, step.y);

            //No tile found? Must have gone over map edge!
            if (! tile)
                return false;

            newRoute.push(this.getDetailsFromTile(tile));
        }

        return true;
    });

    return newRoute;
};

BresenhamRoute.prototype.getDetailsFromTile = function(tile)
{
    return new TileDetailsForRaytrace(
        tile.position,
        tile.concealment,
        tile.subElement,
        tile.elevation,
        tile.elementHeight,
        tile.unique
    );
};

BresenhamRoute.prototype.bresenhamRaytrace = function(start, end)
{
    var x0 = start.x;
    var x1 = end.x;
    var y0 = start.y;
    var y1 = end.y;

    var dx = Math.abs(x1-x0);
    var dy = Math.abs(y1-y0);
    var sx = (x0 < x1) ? 1 : -1;
    var sy = (y0 < y1) ? 1 : -1;
    var err = dx-dy;

    var route = [];

    this.route.push({x:x0, y:y0});

    while(true)
    {
        var e2 = 2*err;

        if (e2 >-dy && e2 < dx)
        {
            this.route.push([{x:x0+sx, y:y0}, {x:x0, y:y0+sy}]);
        }

        if (e2 >-dy)
        {
            err -= dy;
            x0  += sx;
        }

        if (e2 < dx)
        {
            err += dx;
            y0  += sy;
        }

        this.route.push({x:x0, y:y0});

        if ((x0==x1) && (y0==y1))
            break;
    }

    return route;
};

File: app/model/raytracing/BresenhamRoute.scala
package models.raytracing;

import models.repositories.TileRepository
import models.tiles._

class BresenhamRoute(
    val start: (Int, Int),
    val target: (Int, Int),
    val tileRepository: TileRepository)
{
    private lazy val route: List[Array[(Int, Int)]] = bresenham;

    def getRoute: List[Array[(Int, Int)]] = route;

    def getRouteWithTileDetails: List[Array[TileDetailsForRaytrace]] =
    {
        route.map(_.map(tileRepository.getTileDetailsForRaytrace(_)));
    }

    private def bresenham: List[Array[(Int, Int)]] =
    {
        var route: List[Array[(Int, Int)]] = List.empty[Array[(Int, Int)]];

        var (x0, y0) = this.start;
        var (x1, y1) = this.target;

        var dx:Int = math.abs(x1-x0);
        var dy:Int = math.abs(y1-y0);
        var sx:Int = -1;
        if (x0 < x1)
            sx = 1 ;

        var sy:Int = -1;
        if (y0 < y1)
            sy =1;

        var err:Int = dx-dy;

        route :+= Array((x0, y0));

        while((x0 != x1) || (y0 != y1))
        {
            var e2:Int = 2*err;

            if (e2 > -dy && e2 < dx)
            {
                route :+= Array((x0+sx, y0), (x0, y0+sy));
            }

            if (e2 > -dy)
            {
                err -= dy;
                x0  += sx;
            }

            if (e2 < dx)
            {
                err += dx;
                y0  += sy;
            }

            route :+= Array((x0, y0));
        }

        route;
    }
}
